spmd/compiler/fusion.py

```diff
@@ -0,0 +1,182 @@
+from copy import deepcopy
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Iterable, List, Optional
+
+import torch
+import torch.distributed as dist
+import torch.fx as fx
+from .graph_utils import (
+    OP,
+    get_node_tensor_numel,
+    get_output_node,
+    graph_cleanup,
+    pretty_print_graph,
+)
+
+
+# enum for the supported fusion comm types
+class Comm_Type(str, Enum):
+    allreduce = "allreduce_"
+    allgather = "allgather_"
+    broadcast = "broadcast_"
+    reducescatter = "reduce_scatter_"
+    scatter = "scatter_"
+
+
+@dataclass
+class FusionElement:
+    comm_type: Optional[Comm_Type] = None
+    node_list: List = field(default_factory=lambda: [])  # type: ignore
+    size: int = 0
+    prev_node: Optional[fx.Node] = None  # node that was before start of section
+    next_node: Optional[fx.Node] = None  # node that was after end of section
+    processed: bool = False
+    output_name: str = ""
+    clone_node: Optional[fx.Node] = None
+    comm_node: Optional[fx.Node] = None
+    wait_node: Optional[fx.Node] = None
+
+
+@dataclass
+class GraphInfo:
+    len: int = 0
+    global_buffer: Optional[fx.Node] = None
+    global_buffer_size: int = 0
+    output: Optional[fx.Node] = None
+    first: Optional[fx.Node] = None
+
+    def update_info(self, gm: fx.GraphModule, debug: bool = True) -> None:
+        """get the len, input and output nodes"""
+        graph_len = gm.graph._len
+        if not graph_len:
+            raise ValueError("Empty graph passed in....")
+        self.len = graph_len
+
+        nodelist = gm.graph.nodes
+
+        for i, node in enumerate(nodelist):
+            if node.op == OP.PLACEHOLDER:
+                self.first = node
+                break
+
+        self.output = get_output_node(gm)
+        assert (
+            self.output is not None
+        ), f"unable to locate output node in gm {gm.graph}"
+
+        if debug:
+            print(
+                f"updated info - len = {self.len} input = {self.first}, output = {self.output}"
+            )
+
+
+def _insert_buffer_node(
+    gm: fx.GraphModule, insert_before_node: fx.Node, buffer_size: Iterable[int]
+) -> fx.Node:
+    """insert a torch.empty node in front of insert_before_node"""
+    with gm.graph.inserting_before(insert_before_node):
+        new_buffer_node = gm.graph.create_node(
+            OP.CALL_FUNCTION,
+            target=torch.empty,
+            args=tuple(buffer_size),  # TODO - need device
+        )
+    assert (
+        new_buffer_node is not None
+    ), f"failed to create buffer node, size={buffer_size}"
+
+    return new_buffer_node
+
+
+def _scan_graph_for_fusion_elements(
+    gm: fx.GraphModule,
+    comm_type: Comm_Type = Comm_Type.allreduce,
+) -> Optional[List[FusionElement]]:
+    """scan entire graph for matching sections of CommTensor style expansions
+    returns list of FusionElements that match comm_type"""
+
+    element_list = []
+
+    fe_pattern = [
```

Comment:
Can we rename this to `fe_sequence` since the order does matter?

Reply:
fixed